vaadin-spring module validation is failing

Spring 17.0 SNAPSHOT validation is failing

https://bender.vaadin.com/viewLog.html?buildId=224574&tab=buildResultsDiv&buildTypeId=Flow_Flow70_AddOns_Spring170Snapshot

A number of fusion related unit tests are failing.
E.g. 
```

RequestUtilTest.testExternalRequest_init_customMapping
--
Current failure:#861  No changes 15 Apr 21 12:46First failure: #859  No changes 15 Apr 21 11:03java.lang.IllegalStateException: Failed to load ApplicationContext 	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:132) 	at org.springframework.test.context.support.DefaultTestContext.getApplicationContext(DefaultTestContext.java:123) 	at org.springframework.test.context.web.ServletTestExecutionListener.setUpRequestContextIfNecessary(ServletTestExecutionListener.java:190) 	at org.springframework.test.context.web.ServletTestExecutionListener.prepareTestInstance(ServletTestExecutionListener.java:132) 	at org.springframework.test.context.TestContextManager.prepareTestInstance(TestContextManager.java:244) 	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.createTest(SpringJUnit4ClassRunner.java:227) 	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner$1.runReflectiveCall(SpringJUnit4ClassRunner.java:289) 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12) 	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.methodBlock(SpringJUnit4ClassRunner.java:291) 	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:246) 	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:97) 	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331) 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79) 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329) 	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66) 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293) 	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61) 	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70) 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306) 	at org.junit.runners.ParentRunner.run(ParentRunner.java:413) 	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:190) 	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252) 	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141) 	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112) 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) 	at java.lang.reflect.Method.invoke(Method.java:498) 	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189) 	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165) 	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85) 	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115) 	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75) Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'endpointUtil': Unsatisfied dependency expressed through field 'registry'; nested exception is org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type 'com.vaadin.flow.server.connect.EndpointRegistry' available: expected at least 1 bean which qualifies as autowire candidate. Dependency annotations: {@org.springframework.beans.factory.annotation.Autowired(required=true)} 	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:643) 	at org.springframework.beans.factory.annotation.InjectionMetadata.inject(InjectionMetadata.java:119) 	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.postProcessProperties(AutowiredAnnotationBeanPostProcessor.java:399) 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1415) 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:608) 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:531) 	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335) 	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) 	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333) 	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208) 	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:944) 	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:925) 	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:588) 	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:767) 	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:759) 	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:426) 	at org.springframework.boot.SpringApplication.run(SpringApplication.java:326) 	at org.springframework.boot.test.context.SpringBootContextLoader.loadContext(SpringBootContextLoader.java:122) 	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContextInternal(DefaultCacheAwareContextLoaderDelegate.java:99) 	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:124) 	... 32 more Caused by: org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type 'com.vaadin.flow.server.connect.EndpointRegistry' available: expected at least 1 bean which qualifies as autowire candidate. Dependency annotations: {@org.springframework.beans.factory.annotation.Autowired(required=true)} 	at org.springframework.beans.factory.support.DefaultListableBeanFactory.raiseNoMatchingBeanFound(DefaultListableBeanFactory.java:1777) 	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1333) 	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1287) 	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:640) 	... 51 more |   | Current failure: |   | #861 | No changes | 15 Apr 21 12:46 |   | First failure: |   | #859 | No changes | 15 Apr 21 11:03
  | Current failure: |   | #861 | No changes | 15 Apr 21 12:46
  | First failure: |   | #859 | No changes | 15 Apr 21 11:03


```

All failures are : 
```

RequestUtilTest.testExternalRequest_init_customMapping
RequestUtilTest.testExternalRequest_other_customMapping
RequestUtilTest.testRootRequest_init_customMapping
RequestUtilTest.testRootRequest_init_standardMapping
RequestUtilTest.testRootRequest_other_customMapping
RequestUtilTest.testRootRequest_other_standardMapping
RequestUtilTest.testSubRequest_init_customMapping
RequestUtilTest.testSubRequest_init_standardMapping
RequestUtilTest.testSubRequest_other_customMapping
RequestUtilTest.testSubRequest_other_standardMapping
VaadinDefaultRequestCacheTest.endpointRequestNotSaved
VaadinDefaultRequestCacheTest.internalRequestsNotSaved
VaadinDefaultRequestCacheTest.normalRouteRequestSaved
VaadinDefaultRequestCacheTest.serviceWorkerRequestNotSaved

```

--- a/vaadin-spring/src/main/java/com/vaadin/flow/spring/security/RequestUtil.java
+++ b/vaadin-spring/src/main/java/com/vaadin/flow/spring/security/RequestUtil.java
@@ -55,9 +55,31 @@ public class RequestUtil {
         return HandlerHelper.isFrameworkInternalRequest(vaadinMapping, request);
     }
 
+    /**
+     * Checks whether the request targets an endpoint.
+     *
+     * @param request the servlet request
+     * @return {@code true} if the request is targeting an enpoint, {@code false}
+     *         otherwise
+     */
     public boolean isEndpointRequest(HttpServletRequest request) {
         if (endpointUtil != null) {
             return ((EndpointUtil) endpointUtil).isEndpointRequest(request);
+        }
+        return false;
+    }
+
+    /**
+     * Checks whether the request targets an endpoint that is public, i.e. marked
+     * as @{@link com.vaadin.flow.server.connect.auth.AnonymousAllowed}.
+     *
+     * @param request the servlet request
+     * @return {@code true} if the request is targeting an anonymous enpoint,
+     *         {@code false} otherwise
+     */
+    public boolean isAnonymousEndpoint(HttpServletRequest request) {
+        if (endpointUtil != null) {
+            return ((EndpointUtil) endpointUtil).isAnonymousEndpoint(request);
         }
         return false;
     }

--- a/vaadin-spring/src/main/java/com/vaadin/flow/spring/security/VaadinWebSecurityConfigurerAdapter.java
+++ b/vaadin-spring/src/main/java/com/vaadin/flow/spring/security/VaadinWebSecurityConfigurerAdapter.java
@@ -75,6 +75,8 @@ public abstract class VaadinWebSecurityConfigurerAdapter
                 // Vaadin internal requests must always be allowed to allow public Flow pages
                 // and/or login page implemented using Flow.
         urlRegistry.requestMatchers(requestUtil::isFrameworkInternalRequest).permitAll();
+        // Public endpoints are OK to access
+        urlRegistry.requestMatchers(requestUtil::isAnonymousEndpoint).permitAll();
         urlRegistry.requestMatchers(getDefaultHttpSecurityPermitMatcher()).permitAll();
 
         // all other requests require authentication

--- a/vaadin-spring/src/test/java/com/vaadin/flow/spring/security/RequestUtilTest.java
+++ b/vaadin-spring/src/test/java/com/vaadin/flow/spring/security/RequestUtilTest.java
@@ -125,8 +125,7 @@ public class RequestUtilTest {
 
     static HttpServletRequest createRequest(String pathInfo, RequestType type,
             Map<String, String> headers) {
-        String uri = "http://localhost:8080"
-                + (pathInfo == null ? "/" : pathInfo);
+        String uri = (pathInfo == null ? "/" : pathInfo);
         MockHttpServletRequest r = new MockHttpServletRequest("GET", uri);
         r.setPathInfo(pathInfo);
         if (type != null) {

--- a/vaadin-spring/src/test/java/com/vaadin/flow/spring/security/RequestUtilTest.java
+++ b/vaadin-spring/src/test/java/com/vaadin/flow/spring/security/RequestUtilTest.java
@@ -6,8 +6,11 @@ import java.util.Map;
 import javax.servlet.http.HttpServletRequest;
 
 import com.vaadin.flow.server.HandlerHelper.RequestType;
+import com.vaadin.flow.server.connect.EndpointNameChecker;
+import com.vaadin.flow.server.connect.EndpointRegistry;
 import com.vaadin.flow.server.connect.EndpointUtil;
 import com.vaadin.flow.server.connect.VaadinEndpointProperties;
+import com.vaadin.flow.server.connect.auth.VaadinConnectAccessChecker;
 import com.vaadin.flow.shared.ApplicationConstants;
 import com.vaadin.flow.spring.VaadinConfigurationProperties;
 
@@ -23,7 +26,9 @@ import org.springframework.test.context.junit4.SpringRunner;
 
 @RunWith(SpringRunner.class)
 @SpringBootTest(classes = { VaadinDefaultRequestCache.class, RequestUtil.class,
-        EndpointUtil.class, VaadinEndpointProperties.class, })
+        EndpointUtil.class, VaadinEndpointProperties.class,
+        EndpointRegistry.class, EndpointNameChecker.class,
+        VaadinConnectAccessChecker.class })
 public class RequestUtilTest {
 
     @Autowired

--- a/vaadin-spring/src/test/java/com/vaadin/flow/spring/security/VaadinDefaultRequestCacheTest.java
+++ b/vaadin-spring/src/test/java/com/vaadin/flow/spring/security/VaadinDefaultRequestCacheTest.java
@@ -1,13 +1,18 @@
 package com.vaadin.flow.spring.security;
 
+import java.lang.reflect.Method;
 import java.util.Collections;
 
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
 import com.vaadin.flow.server.HandlerHelper.RequestType;
+import com.vaadin.flow.server.connect.Endpoint;
+import com.vaadin.flow.server.connect.EndpointNameChecker;
+import com.vaadin.flow.server.connect.EndpointRegistry;
 import com.vaadin.flow.server.connect.EndpointUtil;
 import com.vaadin.flow.server.connect.VaadinEndpointProperties;
+import com.vaadin.flow.server.connect.auth.VaadinConnectAccessChecker;
 import com.vaadin.flow.spring.VaadinConfigurationProperties;
 
 import org.junit.Assert;
@@ -21,11 +26,14 @@ import org.springframework.test.context.junit4.SpringRunner;
 @RunWith(SpringRunner.class)
 @SpringBootTest(classes = { VaadinDefaultRequestCache.class, RequestUtil.class,
         EndpointUtil.class, VaadinEndpointProperties.class,
-        VaadinConfigurationProperties.class })
+        VaadinConfigurationProperties.class, EndpointRegistry.class,
+        EndpointNameChecker.class, VaadinConnectAccessChecker.class })
 public class VaadinDefaultRequestCacheTest {
 
     @Autowired
     VaadinDefaultRequestCache cache;
+    @Autowired
+    EndpointRegistry endpointRegistry;
     @Autowired
     RequestUtil requestUtil;
 
@@ -61,11 +69,22 @@ public class VaadinDefaultRequestCacheTest {
         Assert.assertNull(cache.getRequest(request, response));
     }
 
+    @Endpoint
+    public static class FakeEndpoint {
+        public void fakeMethod() {
+        }
+    }
+
     @Test
-    public void endpointRequestNotSaved() {
+    public void endpointRequestNotSaved() throws Exception {
         HttpServletRequest request = RequestUtilTest
-                .createRequest("/connect/MyClass/MyEndpoint", null);
+                .createRequest("/connect/fakeendpoint/fakemethod", null);
         HttpServletResponse response = createResponse();
+        Method registerMethod = EndpointRegistry.class
+                .getDeclaredMethod("registerEndpoint", Object.class);
+        registerMethod.setAccessible(true);
+        registerMethod.invoke(endpointRegistry, new FakeEndpoint());
+
         cache.saveRequest(request, response);
         Assert.assertNull(cache.getRequest(request, response));
     }

